pkg/handlers/v1/storage/db.go

```diff
@@ -0,0 +1,309 @@
+
+CREATE TABLE aws_events_ips_hostnames (
+    ts TIMESTAMP NOT NULL,
+    is_public BOOLEAN NOT NULL,
+    is_join BOOLEAN NOT NULL,
+    aws_resources_id VARCHAR NOT NULL,
+    FOREIGN KEY (aws_resources_id) REFERENCES aws_resources (id),
+    aws_ips_ip INET NOT NULL,
+    FOREIGN KEY (aws_ips_ip) REFERENCES aws_ips (ip),
+    aws_hostnames_hostname VARCHAR,
+    FOREIGN KEY (aws_hostnames_hostname) REFERENCES aws_hostnames (hostname))
+PARTITION BY
+    RANGE (
+        ts
+);
+
+-- And we'll make sure there's an index right away:
+
+CREATE INDEX IF NOT EXISTS aws_events_ips_hostnames_aws_ips_ip_ts_idx ON aws_events_ips_hostnames USING BTREE (aws_ips_ip, ts);
+
+Also, some good advice to follow:  https://www.vividcortex.com/blog/2015/09/22/common-pitfalls-go/
+
+*/
+
+const tableAWSResources = "aws_resources"
+const tableAWSIPS = "aws_ips"
+const tableAWSHostnames = "aws_hostnames"
+const tableAWSEventsIPSHostnames = "aws_events_ips_hostnames"
+
+// can't use Sprintf in a const, so...
+const commonQuery = "SELECT " +
+	"	aws_resources_id, aws_ips_ip, aws_hostnames_hostname, is_public, is_join, ts, aws_resources.account_id, aws_resources.region, aws_resources.type, aws_resources.meta " +
+	"FROM " +
+	"	" + tableAWSEventsIPSHostnames + " " +
+	"JOIN " + tableAWSResources + " " +
+	"	ON aws_resources_id = aws_resources.id " +
+	"WHERE "
+
+// DB represents a convenient database abstraction layer
+type DB struct {
+	sqldb *sql.DB      // this is a unit test seam
+	LogFn domain.LogFn // also a seam
+
+	once sync.Once
+}
+
+// Init initializes a connection to a Postgres database according to the environment variables POSTGRES_HOST, POSTGRES_PORT, POSTGRES_USER, POSTGRES_PASSWORD, POSTGRES_DATABASE
+func (db *DB) Init(ctx context.Context, postgresConfig *domain.PostgresConfig) error {
+	var initerr error
+	db.once.Do(func() {
+
+		host := postgresConfig.Hostname
+		port := postgresConfig.Port
+		user := postgresConfig.Username
+		password := postgresConfig.Password
+		dbname := postgresConfig.DatabaseName
+
+		logger := db.LogFn(ctx)
+
+		if db.sqldb == nil {
+			sslmode := "disable"
+			if host != "localhost" {
+				sslmode = "enable"
+			}
+			psqlInfo := fmt.Sprintf("host=%s port=%s user=%s "+
+				"password=%s dbname=%s sslmode=%s",
+				host, port, user, password, dbname, sslmode)
+			pgdb, err := sql.Open("postgres", psqlInfo)
+			if err != nil {
+				logger.Error(logs.DBOpenError{Reason: err.Error()})
+				initerr = err
+			}
+
+			err = pgdb.Ping()
+			if err != nil {
+				logger.Error(logs.DBPingError{Reason: err.Error()})
+				initerr = err
+			}
+
+			db.sqldb = pgdb
+		}
+	})
+	return initerr
+}
+
+// StoreCloudAsset an implementation of the Storage interface that records to a database
+func (db *DB) StoreCloudAsset(ctx context.Context, cloudAssetChanges domain.CloudAssetChanges) error {
+	tx, err := db.sqldb.Begin()
+	if err != nil {
+		return err
+	}
+
+	defer func() {
```

Comment:
if `Commit()` or `Rollback()` return an error, we will miss it. If either return with an error, we'd want to return it from `StoreCloudAsset`. I'd recommend looking at the [example](https://play.golang.org/p/nDjr7Z-wSNR) usage from the golang docs. Though instead of emitting a Fatal log, it may be better to return the original error if the rollback was successful, otherwise [wrap](https://dave.cheney.net/2016/04/27/dont-just-check-errors-handle-them-gracefully) the original error inside the rollback error.

Reply:
OK will improve it